### PR TITLE
new api/op repeat_interleave

### PR DESCRIPTION
**Pull request by kuizhiqing**
### PR types

New features

### PR changes

APIs

### Describe

new api/op repeat_interleave

examples
```
x = paddle.to_tensor([[1, 2, 3], [4, 5, 6]])
repeats  = paddle.to_tensor([3, 2, 1], dtype='int32')

paddle.repeat_interleave(x, repeats, 1)
# [[1, 1, 1, 2, 2, 3],
#  [4, 4, 4, 5, 5, 6]]

paddle.repeat_interleave(x, 2, 0)
# [[1, 2, 3], [1, 2, 3], [4, 5, 6], [4, 5, 6]]

paddle.repeat_interleave(x, 2, None)
# [1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6]
```
 
performance v.s. PyTorch

| input(shape) | output(shape) | paddle(s) | torch(s) |
|---|---|---|---|
|[100, 2, 40]|[100, 7, 40]|6.16|8.20|
|[100, 5, 40]|[100, 29, 40]|6.41|8.34|
|[100,30,40]|[100, 154, 40]|6.97|8.44|
|[100, 50, 40]|[100, 229, 40]|7.64|8.59|
|[100, 100, 40]|[100, 470, 40]|8.88|10.22|
|[100, 1000, 40]|[100, 4589, 40]|31.57|37.41|
|[100, 1000, 40]|[445, 1000, 40]|28.26|35.51|
|[100, 1000, 40]|[100, 1000, 179]|30.05|37.00|

### CODE REVIEW

File: python/paddle/tensor/manipulation.py [hunk 1]
```diff
@@ -2579,4 +2579,58 @@ def as_real(x, name=None):
...
+def repeat_interleave(x, repeats, axis=None, name=None):
+    """
+
+    Returns a new tensor which repeats the ``input`` tensor along dimension ``axis`` using
+    the entries in ``repeats`` which is a int or a Tensor.
+
+    Args:
+        x (Tensor): The input Tensor to be operated. The data of ``x`` can be one of float32, float64, int32, int64.
+        repeats (Tensor or int): The repetitions times of each element.
+        axis (int, optional): The dimension in which we manipulate. Default: if None, the output tensor ss flatten.
```
Comment (qingqing01) [inline python/paddle/tensor/manipulation.py]: ss -> is ?

File: python/paddle/tensor/manipulation.py [hunk 2]
```diff
@@ -2579,4 +2579,58 @@ def as_real(x, name=None):
         dtype=_complex_to_real_dtype(x.dtype))
     outputs = {"Out": out}
     helper.append_op(type=op_type, inputs=inputs, outputs=outputs)
+
+
+def repeat_interleave(x, repeats, axis=None, name=None):
+    """
+
+    Returns a new tensor which repeats the ``input`` tensor along dimension ``axis`` using
```
Comment (qingqing01) [inline python/paddle/tensor/manipulation.py]: input - >  x

File: paddle/fluid/operators/repeat_interleave_op.cu [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        auto* out_data = out->mutable_data<T>(context.GetPlace());
+        int64_t numel = out->numel();
+        int64_t size = output_dim[dim];
+        int64_t delta = input_dim[dim] - size;
+
+        index_select_cuda_kernel<T, int64_t><<<
+            (numel + PADDLE_CUDA_NUM_THREADS - 1) / PADDLE_CUDA_NUM_THREADS,
+            PADDLE_CUDA_NUM_THREADS, 0, stream>>>(in_data, out_data, index_data,
+                                                  numel, stride, size, delta);
+        platform::GpuStreamSync(stream);
```
Comment (qingqing01) [inline paddle/fluid/operators/repeat_interleave_op.cu]: There is no need to sync  stream here

File: paddle/fluid/operators/repeat_interleave_op.cu [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+              "Input(RepeatsTensor) holds the wrong type, it holds %s, but "
+              "desires to be %s or %s",
+              paddle::framework::DataTypeToString(index_type),
+              paddle::framework::DataTypeToString(
+                  framework::proto::VarType::INT32),
+              paddle::framework::DataTypeToString(
+                  framework::proto::VarType::INT64)));
+
+      if (index_type == framework::proto::VarType::INT64) {
+        RepeatsTensor2IndexTensor<DeviceContext, int64_t>(*repeats_tensor,
```
Comment (qingqing01) [inline paddle/fluid/operators/repeat_interleave_op.cu:117]: This is a CPU kernel, which will affect performance

File: python/paddle/fluid/tests/unittests/test_repeat_interleave_op.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+from __future__ import print_function
+
+import unittest
+import paddle
+import numpy as np
+import paddle.fluid.core as core
+import paddle.fluid as fluid
+from paddle.fluid import Program, program_guard
+
+paddle.set_device('cpu')
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_repeat_interleave_op.py]: 1. Not test GPU? 
2. Need to test gradients

File: python/paddle/__init__.py [hunk 6]
```diff
@@ -159,6 +159,7 @@
 from .tensor.manipulation import as_complex  # noqa: F401
 from .tensor.manipulation import as_real  # noqa: F401
 from .tensor.manipulation import moveaxis  # noqa: F401
+from .tensor.manipulation import repeat_interleave  # noqa: F401
```
Comment (TCChenlong) [inline python/paddle/__init__.py:162]: 这里还要将API加在 __all__ 里；

File: python/paddle/tensor/__init__.py [hunk 7]
```diff
@@ -114,6 +114,7 @@
 from .manipulation import as_complex  # noqa: F401
 from .manipulation import as_real  # noqa: F401
 from .manipulation import moveaxis  # noqa: F401
+from .manipulation import repeat_interleave  # noqa: F401
```
Comment (TCChenlong) [inline python/paddle/tensor/__init__.py:117]: 如果支持 Tensor.repeat_interleave() 这样的用法，也需要在 tensor_method_func 这里添加

File: python/paddle/tensor/manipulation.py [hunk 8]
```diff
@@ -2582,6 +2582,63 @@ def as_real(x, name=None):
...
+            An int repeats will repeat every elements while a Tensor repeats with the same length indicate 
+            the repetitions of each element with the same index.
+        axis (int, optional): The dimension in which we manipulate. Default: if None, the output tensor is flatten.
+        name(str, optional): The default value is None. Normally there is no
+            need for user to set this property. For more information, please
+            refer to :ref:`api_guide_Name`.
+
+    Returns:
+        Tensor: A Tensor with same data type as ``x``.
+
```
Comment (TCChenlong) [inline python/paddle/tensor/manipulation.py:2601]: 示例代码格式有问题。
示例代码建议提供3个:
1：repeats = int 如：(2)
2: repeats = Tensor 如：([3, 2, 1])
3、axis = None；
这样会更清楚一些

File: python/paddle/tensor/manipulation.py [hunk 9]
```diff
@@ -2582,6 +2582,63 @@ def as_real(x, name=None):
...
 
+def repeat_interleave(x, repeats, axis=None, name=None):
+    """
+
+    Returns a new tensor which repeats the ``x`` tensor along dimension ``axis`` using
+    the entries in ``repeats`` which is a int or a Tensor.
+
+    Args:
+        x (Tensor): The input Tensor to be operated. The data of ``x`` can be one of float32, float64, int32, int64.
+        repeats (Tensor or int): The repetitions times of each element along the given axis, 0 means no repeat.
```
Comment (TCChenlong) [inline python/paddle/tensor/manipulation.py]: repeats 为 Tensor 的时候，建议单独写一句。现在写的太长了，不好理解。

File: paddle/fluid/operators/repeat_interleave_op.cc [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    AddInput("X", "(Tensor) the input tensor.");
+    AddInput("RepeatsTensor",
+             "the 1-D tensor containing the repeats alongsize the axis.")
+        .AsDispensable();
+    AddOutput("Out", "the output tensor.");
+    AddAttr<int>("Repeats", "the number of repetitions for each element.")
+        .SetDefault(0);
+    AddAttr<int>("dim", "the dimension in which we repeat.").SetDefault(0);
+    AddComment(R"DOC(
+    Returns a new tensor which repeats the input tensor
```
Comment (xymyeah) [inline paddle/fluid/operators/repeat_interleave_op.cc]: AddComment的格式可能需要修改一下，可以参考一下https://github.com/PaddlePaddle/docs/blob/release/1.2/doc/fluid/dev/name_convention.md

File: paddle/fluid/operators/repeat_interleave_op.cc [hunk 11]
```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (xymyeah) [inline paddle/fluid/operators/repeat_interleave_op.cc]: Copyright (c) 2021 ？

File: python/paddle/fluid/tests/unittests/test_repeat_interleave_op.py [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+                dtype='int32',
+                append_batch_size=False)
+            z = paddle.repeat_interleave(x, index, axis=1)
+            exe = fluid.Executor(fluid.CPUPlace())
+            res, = exe.run(feed={'x': self.data_x,
+                                 'repeats': self.data_index},
+                           fetch_list=[z.name],
+                           return_numpy=False)
+        expect_out = np.array([[2.0, 3.0, 3.0, 4.0], [6.0, 7.0, 7.0, 8.0],
+                               [10.0, 11.0, 11.0, 12.0]])
```
Comment (jeff41404) [inline python/paddle/fluid/tests/unittests/test_repeat_interleave_op.py]: it's better to use result of numpy.repeat, not just fix it
the same as below

File: python/paddle/fluid/tests/unittests/test_repeat_interleave_op.py [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+        # case axis none
+        with fluid.dygraph.guard():
+            x = fluid.dygraph.to_variable(
+                np.array([[1, 2, 1], [1, 2, 3]]).astype('int32'))
+            index = fluid.dygraph.to_variable(
+                np.array([1, 1, 2, 1, 2, 2]).astype('int32'))
+            z = paddle.repeat_interleave(x, index, None)
+            np_z = z.numpy()
+        expect_out = np.array([1, 2, 1, 1, 1, 2, 2, 3, 3])
+        self.assertTrue(np.allclose(expect_out, np_z))
```
Comment (jeff41404) [inline python/paddle/fluid/tests/unittests/test_repeat_interleave_op.py]: it's better to use result of numpy.repeat, not just fix it
the same as below

File: paddle/fluid/operators/repeat_interleave_op.h [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+        RepeatsTensor2IndexTensor<DeviceContext, int64_t>(*repeats_tensor,
+                                                          &index);
+        auto output_dim = framework::vectorize(inputs.dims());
+        output_dim[dim] = index.dims()[0];
+        output->Resize(framework::make_ddim(output_dim));
+        IndexSelectInner<DeviceContext, T, int64_t>(context, &inputs, index,
+                                                    output, dim);
+      }
+    } else if (repeats > 0) {
+      int index_size = inputs.dims()[dim] * repeats;
```
Comment (jeff41404) [inline paddle/fluid/operators/repeat_interleave_op.h]: inputs.dims()[dim] will return data type of int64_t, index_size should be define as int64_t ?

File: paddle/fluid/operators/repeat_interleave_op.h [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+        IndexSelectGradInner<DeviceContext, T, int>(context, *out_grad, index,
+                                                    x_grad, dim);
+      } else if (index_type == framework::proto::VarType::INT64) {
+        RepeatsTensor2IndexTensor<DeviceContext, int64_t>(*repeats_tensor,
+                                                          &index);
+        IndexSelectGradInner<DeviceContext, T, int64_t>(context, *out_grad,
+                                                        index, x_grad, dim);
+      }
+    } else if (repeats > 0) {
+      int index_size = x_grad->dims()[dim] * repeats;
```
Comment (jeff41404) [inline paddle/fluid/operators/repeat_interleave_op.h]: x_grad->dims()[dim] will return data type of int64_t, index_size should be define as int64_t ?

File: paddle/fluid/operators/repeat_interleave_op.cu [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+        const int* index_data = index.data<int>();
+        index_select_grad_cuda_kernel<T, int><<<
+            (out_nums + PADDLE_CUDA_NUM_THREADS - 1) / PADDLE_CUDA_NUM_THREADS,
+            PADDLE_CUDA_NUM_THREADS, 0, stream>>>(
+            output_grad_data, in_grad_data, index_data, index_nums, out_nums,
+            stride, size, delta);
+        platform::GpuStreamSync(stream);
+      }
+    } else if (repeats > 0) {
+      int index_size = in_grad->dims()[dim] * repeats;
```
Comment (jeff41404) [inline paddle/fluid/operators/repeat_interleave_op.cu]: in_grad->dims()[dim] will return data type of int64_t, index_size should be define as int64_t ?

File: paddle/fluid/operators/repeat_interleave_op.cu [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+        int64_t size = output_dim[dim];
+        int64_t delta = input_dim[dim] - size;
+
+        index_select_cuda_kernel<T, int><<<
+            (numel + PADDLE_CUDA_NUM_THREADS - 1) / PADDLE_CUDA_NUM_THREADS,
+            PADDLE_CUDA_NUM_THREADS, 0, stream>>>(in_data, out_data, index_data,
+                                                  numel, stride, size, delta);
+      }
+    } else if (repeats > 0) {
+      int index_size = in->dims()[dim] * repeats;
```
Comment (jeff41404) [inline paddle/fluid/operators/repeat_interleave_op.cu]: in->dims()[dim] will return data type of int64_t, index_size should be define as int64_t ?

**Comment by jeff41404** [comment]
shall we add example of API in Describe ?

File: python/paddle/tensor/manipulation.py [hunk 18]
```diff
@@ -2582,6 +2582,68 @@ def as_real(x, name=None):
...
+    """
+
+    Returns a new tensor which repeats the ``x`` tensor along dimension ``axis`` using
+    the entries in ``repeats`` which is a int or a Tensor.
+
+    Args:
+        x (Tensor): The input Tensor to be operated. The data of ``x`` can be one of float32, float64, int32, int64.
+        repeats (Tensor or int): The number of repetitions for each element. repeats is broadcasted to fit the shape of the given axis.
+        axis (int, optional): The dimension in which we manipulate. Default: if None, the output tensor is flatten.
+        name(str, optional): The default value is None. Normally there is no
```
Comment (TCChenlong) [inline python/paddle/tensor/manipulation.py:2595]: if None -> None

File: python/paddle/tensor/manipulation.py [hunk 19]
```diff
@@ -2582,6 +2582,68 @@ def as_real(x, name=None):
...
+            need for user to set this property. For more information, please
+            refer to :ref:`api_guide_Name`.
+
+    Returns:
+        Tensor: A Tensor with same data type as ``x``.
+
```
Comment (TCChenlong) [inline python/paddle/tensor/manipulation.py:2601]: 示例代码要显式的写 Examples：

**Review by jeff41404 — APPROVED** [review]
LGTM

File: paddle/fluid/operators/repeat_interleave_op.cu [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+                                                          &index);
+        int64_t index_nums = index.numel();
+
+        const int64_t* index_data = index.data<int64_t>();
+        index_select_grad_cuda_kernel<T, int64_t><<<
+            (out_nums + PADDLE_CUDA_NUM_THREADS - 1) / PADDLE_CUDA_NUM_THREADS,
+            PADDLE_CUDA_NUM_THREADS, 0, stream>>>(
+            output_grad_data, in_grad_data, index_data, index_nums, out_nums,
+            stride, size, delta);
+        platform::GpuStreamSync(stream);
```
Comment (xymyeah) [inline paddle/fluid/operators/repeat_interleave_op.cu:243]: platform::GpuStreamSync(stream)是否是必须的？其他地方也类似

**Review by xymyeah — COMMENTED** [review]
LGTM

File: paddle/fluid/operators/repeat_interleave_op.cu (hunk 20 above)
Comment (kuizhiqing) [reply paddle/fluid/operators/repeat_interleave_op.cu:243]: 这里可以保留

**Review by Superjomn — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]
TODO: Fix API Docs

**Review by ForFishes — APPROVED** [review]
LGTM

**Pull request merged**